packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2839,33 +2839,84 @@ export class ContainerRuntime
 		runtimeBatch: boolean,
 	) {
 		if (locationInBatch.batchStart) {
-			const firstMessage = messages[0]?.message;
+			const firstMessage = messagesWithMetadata[0]?.message;
 			assert(firstMessage !== undefined, 0xa31 /* Batch must have at least one message */);
 			this.scheduleManager.batchBegin(firstMessage);
 		}
 
 		let error: unknown;
 		try {
-			messages.forEach(({ message, localOpMetadata }) => {
-				this.ensureNoDataModelChanges(() => {
-					if (runtimeBatch) {
-						this.validateAndProcessRuntimeMessage({
-							message: message as InboundSequencedContainerRuntimeMessage,
-							local,
-							savedOp,
-							localOpMetadata,
-						});
-					} else {
+			if (!runtimeBatch) {
+				// If this is not a runtime message, we need to process it here.
+				messagesWithMetadata.forEach(({ message }) => {
+					this.ensureNoDataModelChanges(() => {
 						this.observeNonRuntimeMessage(message);
+					});
+				});
+				return;
+			}
+
+			/**
+			 * For runtime messages, bunch contiguous messages of the same type and process them together.
+			 * This is an optimization mainly for DDSes, where it can process a bunch of ops together. DDSes
+			 * like merge tree or shared tree can process ops more efficiently when they are bunched together.
+			 */
+			let bunchedOps: {
```

Comment:
Nit: bothers me that we are using `messages` terminology everywhere, but just this one variable is called `bunchedOps`.

Reply:
This code has changed in the last iteration.